src/lava/utils/profiler.py

```diff
@@ -12,3 +12,125 @@
+        self.buffer_size = buffer_size
+        self.used_resources: ty.List[AbstractComputeResource] = []
+
+    def profile(self, proc: AbstractProcess):
+        proc.run = types.MethodType(self.run, proc)
+
+    def get_energy(self) -> np.array:
+        """Returns the energy estimate per time step in µJ."""
+        ...
+        return 0
+
+    def get_power(self) -> np.array:
+        """Returns the power estimate per time step in µW."""
+        ...
+        return 0
+
+    def get_execution_time(self) -> np.array:
+        """Returns the execution time estimate per time step in µs."""
+        ...
+        return 0
+
+    def run(self, proc: AbstractProcess, condition: AbstractRunCondition = None,
+            run_cfg: RunConfig = None):
+        """Runs process given RunConfig and RunCondition.
+
+        Functionally, this method does the same as run(..) of AbstractProcess,
+        but modifies the chosen ProcModels and executables to be able to use the
+        LavaProfiler. From the user perspective, it should not be noticeable as
+        the API does not change. This method will be used to override the method
+        run(..) of an instance of AbstractProcess, when the LavaProfiler is
+        created.
+
+        Parameters
+        ----------
+        proc : AbstractProcess
+            Process instance which run(..) was initially called on.
+        condition : AbstractRunCondition
+            RunCondition instance specifies for how long to run the process.
+        run_cfg : RunConfig
+            RunConfig is used by compiler to select a ProcessModel for each
+            compiled process.
+        """
+
+        if not proc._runtime:
+
+            compiler = Compiler(loglevel=proc.loglevel)
+            # initializer = Initializer()
+
+            # 1. get proc_map
+            # proc_map = initializer._map_proc_to_model(proc, run_cfg)
+            proc_map = compiler._map_proc_to_model(
+                compiler._find_processes(proc), run_cfg)
+
+            # 2. modify proc_map
+            proc_map = self._modify_proc_map(proc_map)
+
+            # 3. set up profiler
+            self._set_profiler(proc_map)
+
+            # 4. create executable
+            executable = compiler.compile(proc, run_cfg)
+
+            # 5. append profiler sync channels
+            self._set_profiler_sync_channel_builders(executable)
+
+            # 6. create Runtime
+            proc._runtime = Runtime(executable,
+                                    ActorType.MultiProcessing,
+                                    loglevel=proc.loglevel)
+            proc._runtime.initialize()
+
+        proc._runtime.start(condition)
+
+    def _modify_proc_map(self, proc_map):
+        """Check if chosen process models have a profileable version and
+        exchange the process models accordingly.
+        Tell the user which Processes will not be profiled, as they lack a
+        profileable ProcModel."""
+        ...
```

Comment:
Should we automatically switch process models? I could imagine that people may get confused in some cases. Let's assume the user first runs a process without profiler and the compiler chooses a process model that is not profilable. Then the user runs the same process with profiler. The profiler will automatically switch the process model to one that can be profiled. But these process models may differ - maybe because of a bug, maybe because of other reasons. The user will not have expected to see any different process behavior just because (s)he activated the profiler. Instead, I would expect that I receive an error message if the default process model cannot be profiled.

Reply:
My approach would be that each ProcModel has a profileable counter part ProcModel, which only adds the operation counters.
So for the LIF neuron for example, we would need a profilable ProcModel for "fixed_pt" and "floating_pt" and there is a 1:1 mapping to exchange them if the profiler is present.

If there is no such ProcModel, then the user will be informed that this Process is not considered by the Profiler. If no chosen ProcModel has a profileable version and we run on simulation only, than there will be an error stating no Process is considered for the Profiler.

Reply:
What does this mean in terms of code duplication? Does it mean there is a ProcModel class and then an almost identical ProcModel class that just adds counters?

Before we go into a whole lot of implementation, as usual, we should first write down an end to end (mock) example of what we are trying to enable and then agree that this is the best way to go. Such decisions are best made not in the abstract, for people who have not thought about the pros and cons deeply before, but using a concrete exmple.

Do we have such an example already?

If not I suggest you draft one and share it. We are at an important fork in the road, so we should get some wider input.

Reply:
I added an example in proc/lif/models.py

We can inherit the ProcModel and add the code for operation counters.